src/dbnode/storage/index/aggregate_iter.go

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"time"
+
+	"github.com/m3db/m3/src/m3ninx/index/segment"
+	"github.com/m3db/m3/src/x/context"
+	xerrors "github.com/m3db/m3/src/x/errors"
+)
+
+var _ AggregateIterator = &aggregateIter{}
+
+type aggregateIter struct {
+	// immutable state
+	readers     []segment.Reader
+	iterateOpts fieldsAndTermsIteratorOpts
+	newIterFn   newFieldsAndTermsIteratorFn
+
+	// immutable state after first Next() call
+	iters          []fieldsAndTermsIterator
+	searchDuration time.Duration
+
+	// mutable state
+	idx                 int
+	err                 error
+	done                bool
+	currField, currTerm []byte
+	nextField, nextTerm []byte
+}
+
+func (it *aggregateIter) Next(ctx context.Context) bool {
+	if it.Done() {
+		return false
+	}
+	if it.iters == nil {
+		for _, reader := range it.readers {
+			iter, err := it.newIterFn(ctx, reader, it.iterateOpts)
+			if err != nil {
+				it.err = err
+				return false
+			}
+			it.searchDuration += iter.SearchDuration()
+			it.iters = append(it.iters, iter)
+		}
+		if !it.next() {
+			it.done = true
+			return false
+		}
+		it.nextField, it.nextTerm = it.current()
+	}
+	// the fieldAndTermsIterator mutates the underlying byte slice, so we need to copy to preserve the value.
+	it.currField = append(it.currField[:0], it.nextField...)
+	it.currTerm = append(it.currTerm[:0], it.nextTerm...)
+
+	if it.next() {
+		it.nextField, it.nextTerm = it.current()
+	} else {
+		it.done = true
```

Comment:
Shouldn't we return false here if `done = true`?

Reply:
no since this is a look ahead iterator. `done` means that the `next*` fields are not set, but the `curr*` fields are set, so the method needs to return true, so they can be read by `Current`. 

should probably add some comments about this. also should add some tests